extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,216 @@
+  ValidAdContainerTypes,
+} from '../../../ads/google/a4a/utils';
+import {user} from '../../../src/log';
+
+/**
+ * - visibilePercentageMin: The percentage of pixels that need to be on screen
+ *   for the creative to be considered "visible".
+ * - totalTimeMin: The total amount of time, in milliseconds, that the creative
+ *   must be on screen in order to be considered "visible".
+ * - continuousTimeMin: The amount of continuous time, in milliseconds, that
+ *   the creative must be on screen for in oreder to be considered "visible".
+ *
+ * @typedef {{
+ *   visiblePercentageMin: number,
+ *   totalTimeMin: number,
+ *   continuousTimeMin: number,
+ * }}
+ */
+export let RefreshConfig;
+
+export const MIN_REFRESH_INTERVAL = 30;
+export const DATA_ATTR_NAME = 'data-enable-refresh';
+export const METATAG_NAME = 'amp-ad-enable-refresh';
+
+const TAG = 'AMP-AD';
+
+export class RefreshManager {
+
+  /**
+   * @param {!./amp-a4a.AmpA4A} a4a The AmpA4A instance to be refreshed.
+   * @param {!RefreshConfig} config
+   */
+  constructor(a4a, config) {
+
+    /** @const @private {!./amp-a4a.AmpA4A} */
+    this.a4a_ = a4a;
+
+    /** @const @private {!Window} */
+    this.win_ = a4a.win;
+
+    /** @const @private {!Element} */
+    this.element_ = a4a.element;
+
+    /** @const @private {string} */
+    this.adType_ = this.element_.getAttribute('type').toLowerCase();
+
+    /** @const @private {?number} */
+    this.refreshInterval_ = this.getPublisherSpecifiedRefreshInterval_();
+
+    /** @const @private {!RefreshConfig} */
+    this.config_ = this.convertConfiguration_(config);
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = isExperimentOn(this.win_, 'amp-ad-refresh') &&
+        // The network has opted-in.
+        !!(this.config_
+        // The publisher has enabled refresh on this slot.
+        && (this.refreshInterval_ || this.refreshInterval_ == '')
+        // The slot is contained only within container types eligible for
+        // refresh.
+        && !getEnclosingContainerTypes(this.element_).filter(container =>
+          container != ValidAdContainerTypes['AMP-CAROUSEL']
+          && container != ValidAdContainerTypes['AMP-STICKY-AD']).length);
+
+    if (this.isRefreshable_) {
+      this.initiateRefreshCycle();
+    }
+  }
+
+
+  /**
+   * Initiates the refresh cycle by initiating the visibility manager on the
+   * element.
+   *
+   * @return {!Promise<boolean>} Promise that resolves to true if refresh event
+   *    completes successfully and false otherwise. This is particularly useful
+   *    for testing.
+   */
+  initiateRefreshCycle() {
+    if (!this.isRefreshable()) {
+      return Promise.resolve(false);
+    }
+    return new Promise(resolve => {
+      analyticsForDoc(this.element_, true).then(analytics => {
```

Comment:
There will be the hard question where do we put the code if we want to share.
We certainly don't want it to be in amphtml/core to bloat the size.
We also don't want a4a to have direct code dependency with analytics.

And also, given that your logic might be simpler than the visibilitymanager in amp-analytics, implementing your own might not be bad idea. You can still reuse the lower level IntersectionObserver code (polyfill etc).